x/profiles/types/msgs.go

```diff
@@ -486,3 +488,128 @@ func (msg MsgUnblockUser) GetSigners() []sdk.AccAddress {
 	blocker, _ := sdk.AccAddressFromBech32(msg.Blocker)
 	return []sdk.AccAddress{blocker}
 }
+
+// ___________________________________________________________________________________________________________________
+
+func NewMsgLinkChainAccount(
+	sourceAddress AddressData,
+	sourceProof Proof,
+	sourceChainConfig ChainConfig,
+	destinationAddress string,
+	destinationProof Proof,
+) *MsgLinkChainAccount {
+	addressAny, err := codectypes.NewAnyWithValue(sourceAddress)
+	if err != nil {
+		panic("failed to pack public key to any type")
+	}
+	return &MsgLinkChainAccount{
+		SourceAddress:      addressAny,
+		SourceProof:        sourceProof,
+		SourceChainConfig:  sourceChainConfig,
+		DestinationAddress: destinationAddress,
+		DestinationProof:   destinationProof,
+	}
+}
+
+// Route should return the name of the module
+func (msg MsgLinkChainAccount) Route() string { return RouterKey }
+
+// Type should return the action
+func (msg MsgLinkChainAccount) Type() string {
+	return ActionLinkChainAccount
+}
+
+// ValidateBasic runs stateless checks on the message
+func (msg MsgLinkChainAccount) ValidateBasic() error {
+	if msg.SourceAddress == nil {
+		return fmt.Errorf("source address cannot be nil")
+	}
+	if err := msg.SourceProof.Validate(); err != nil {
+		return err
+	}
+	if err := msg.SourceChainConfig.Validate(); err != nil {
+		return err
+	}
+	if _, err := sdk.AccAddressFromBech32(msg.DestinationAddress); err != nil {
+		return fmt.Errorf("invalid destination address: %s", msg.DestinationAddress)
+	}
+	if err := msg.DestinationProof.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetSignBytes encodes the message for signing
+func (msg MsgLinkChainAccount) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&msg))
+}
+
+// UnpackInterfaces implements codectypes.UnpackInterfacesMessage
+func (msg *MsgLinkChainAccount) UnpackInterfaces(unpacker codectypes.AnyUnpacker) error {
+	var address AddressData
+	if err := unpacker.UnpackAny(msg.SourceAddress, &address); err != nil {
+		return err
+	}
+
+	var pubKey cryptotypes.PubKey
+	if err := unpacker.UnpackAny(msg.SourceProof.PubKey, &pubKey); err != nil {
+		return err
+	}
+	if err := unpacker.UnpackAny(msg.DestinationProof.PubKey, &pubKey); err != nil {
+		return err
+	}
```

Comment:
There's no need to do this by hand since if the field types implement `UnpackInterfaces`, this method will call them as well automatically.

Reply:
If removing it, the msg `GetSignBytes()` will occurs panic, it said:
```
JSON marshal marshaling error for &Any{TypeUrl:/cosmos.crypto.secp256k1.PubKey,Value:[10 33 3 187 247 28 72 226 52 237 163 92 144 62 218 7 1 180 237 149 67 165 108 138 187 59 24 224 243 237 211 119 244 126 208],XXX_unrecognized:[],}, 
this is likely because amino is being used directly (instead of codec.LegacyAmino which is preferred) 
or UnpackInterfacesMessage is not defined for some type which contains a protobuf Any either directly or via one of its members.
```

Reply:
I think this is related to this: https://github.com/desmos-labs/desmos/pull/449#discussion_r644591486. We should implement the `UnpackInterfacesMessage` for the `Proof` object directly. Then it will be called recursively on it

Reply:
@RiccardoM The panic happens after implementing `UnpackInterfacesMessage ` for the Proof. `GetSignBytes()` calls `MustMarshalJSON()` function to deserialize the proto message. Not sure if it calls `UnpackInterfacesMessage ` as well, let me check it.

Reply:
@dadamu Can you reproduce this on a test please? I would like to take a look

Reply:
@RiccardoM 
It seems that `UnpackInterfacesMessage` doesn't call the child property's `UnpackInterface` recursively. The interface code is [here](https://github.com/cosmos/cosmos-sdk/blob/master/codec/types/interface_registry.go#L60).

I think after implementing `UnpackInterface` for Proof, it can be:
```go
func (msg *MsgLinkChainAccount) UnpackInterfaces(unpacker codectypes.AnyUnpacker) error {
	var address AddressData
	if err := unpacker.UnpackAny(msg.SourceAddress, &address); err != nil {
		return err
	}
	if err := msg.SourceProof.UnpackInterfaces(unpacker); err != nil {
		return err
	}
	if err := msg.DestinationProof.UnpackInterfaces(unpacker); err != nil {
		return err
	}
	return nil
}
```

Reply:
@RiccardoM 
I create a test branch [here](https://github.com/dadamu/desmos/tree/test/proof-unpack).

Reply:
@dadamu You're right, we should call the `UnpackInterfaces` for the nested fields manually using the code you have written. It's also specified [here](https://github.com/cosmos/cosmos-sdk/blob/master/codec/types/interface_registry.go#L71) as you pointed out